File: labbox_ephys_widgets/widgets/SortingsView/SortingsView.py
from .lbdb import find, remove

class SortingsView:
    def __init__(self):
        super().__init__()

    def javascript_state_changed(self, prev_state, state):
        self._refresh_sortings()

    def on_message(self, msg):
        if msg['action'] == 'remove_sortings':
            sorting_ids = msg['sorting_ids']
            remove(collection='sortings', query={'sorting_id': {'$in': sorting_ids}})
            self._refresh_sortings()
        elif msg['action'] == 'refresh_sortings':
            self._refresh_sortings()
    
    def _refresh_sortings(self):
        self._set_status('running', 'Loading sortings...')
        self._set_state(sortings=None)
        sortings = find(collection='sortings', query=dict())
        self._set_state(sortings=sortings)
        self._set_status('finished', f'Finished loading {len(sortings)} sortings.')
    
    # Send a custom message to JavaScript side
    # In .js file, use this.pythonInterface.onMessage((msg) => {...})
    def _send_message(self, msg):
        self.send_message(msg)

    # Set the python state
    def _set_state(self, **kwargs):
        self.set_state(kwargs)
    
    # Set error status with a message
    def _set_error(self, error_message):
        self._set_status('error', error_message)
    
    # Set status and a status message. Use running', 'finished', 'error'
    def _set_status(self, status, status_message=''):
        self._set_state(status=status, status_message=status_message)

File: labbox_ephys_widgets/widgets/SortingsView/SortingsView.js
import React, { Component } from 'react';
import { PythonInterface } from 'reactopya';
import SortingsTable from './SortingsTable.js'
import { Toolbar, IconButton } from '@material-ui/core';
import { FaTrash, FaSync } from 'react-icons/fa';
const config = require('./SortingsView.json');

export default class SortingsView extends Component {
    static title = 'View database of sortings'
    static reactopyaConfig = config
    constructor(props) {
        super(props);
        this.state = {
            // javascript state
            
            // python state
            sortings: null,
            status: '',
            status_message: '',

            //
            selectedSortingIds: {}
        }
    }
    componentDidMount() {
        this.pythonInterface = new PythonInterface(this, config);
        this.pythonInterface.start();
        this.setState({
            status: 'started',
            status_message: 'Starting python backend'
        });
        this.pythonInterface.setState({
            trigger: 1
        });
    }
    componentWillUnmount() {
        this.pythonInterface.stop();
    }
    _handleDeleteSelectedSortings = () => {
        const selectedSortingIds = this.state.selectedSortingIds;
        this.pythonInterface.sendMessage({action: 'remove_sortings', sorting_ids: Object.keys(selectedSortingIds)});
    }
    _handleRefreshSortings = () => {
        this.setState({
            sortings: null,
            status: '',
            status_message: ''
        });
        this.pythonInterface.sendMessage({action: 'refresh_sortings'});
    }
    render() {
        const sortings = this.state.sortings;
        const selectedSortingIds = this.state.selectedSortingIds;

        let content;
        if (sortings) {
            content = (
                <SortingsTable
                    sortings={sortings}
                    selectedSortingIds={selectedSortingIds}
                    onSelectedSortingIdsChanged={(ids) => {this.setState({selectedSortingIds: ids})}}
                />
            )
        }
        else {
            content = (
                <ReportStatus {...this.state} />
            );
        }

        return (
            <div>
                <Toolbar>
                    <IconButton title={"Refresh sortings"} onClick={() => {this._handleRefreshSortings()}}>
                        <FaSync />
                    </IconButton>
                    <IconButton disabled={isEmpty(selectedSortingIds)} title={"Delete selected sortings"} onClick={() => {this._handleDeleteSelectedSortings()}}>
                        <FaTrash />
                    </IconButton>
                </Toolbar>
                <div style={{height: 250, overflowY: 'auto'}}>
                    {content}
                </div>
            </div>
        );
    }
}

class ReportStatus extends Component {
    state = {}
    render() {
        switch (this.props.status) {
            case 'started':
                return <div>Started: {this.props.status_message}</div>;
            case 'running':
                return <div>{this.props.status_message}</div>;
            case 'error':
                return <div>Error: {this.props.status_message}</div>;
            case 'finished':
                return <div>Finished: {this.props.status_message}</div>;
            default:
                return <div>Unknown status: {this.props.status}</div>;
        }
    }
}

function isEmpty(obj) {
    return (Object.getOwnPropertyNames(obj).length == 0);
}